torch/distributed/pipelining/stage.py

```diff
@@ -1293,22 +1314,102 @@ def stage_global_rank(peer_rank):
-            f"output: {[output.shape for output in self.get_outputs_meta()]}"
+            f"inputs: {[inp.shape for inp in self.inputs_meta] if self.inputs_meta is not None else None}, "
+            f"output: {[output.shape for output in self.get_outputs_meta()] if self.inputs_meta is not None else None}"
+        )
+
+    def _shape_inference(
+        self,
+        args: Tuple[Any, ...],
+        kwargs: Optional[Dict[str, Any]] = None,
+    ):
+        if kwargs is None:
+            kwargs = {}
+        assert args is not None, "Args may be an empty list but not None"
+        if (
+            self.is_first
+            or self.stage_index_to_group_rank[self.stage_index - 1] == self.group_rank
+        ):
+            args = tree_map_only(torch.Tensor, lambda x: x.to("meta"), args)
+        else:
+            assert (
+                len(args) == 0
+            ), "Can't supply input args for shape inference on non-first stage"
+            objects = [None]
+            logger.debug(
+                "%s receiving from stage %s", self.stage_index, self.stage_index - 1
+            )
+            dist.recv_object_list(
+                objects, src=self.prev_stage, group=self.group, device=self.device
+            )
+            recv_args = objects[0]
+            assert isinstance(recv_args, tuple), type(recv_args)
+            args = recv_args
+
+        # cache input shapes for use during recv buffer allocation
+        self.inputs_meta = args
+        args = tree_map_only(
+            torch.Tensor, lambda x: torch.zeros_like(x, device=self.device), args
+        )
+
+        # set attributes needed for forward
+        with torch.no_grad():
+            logger.debug("%s running forward", self.stage_index)
+            outputs = self.submod(*args, **kwargs)
+
+        # if single tensor, convert so it is always a list
+        if isinstance(outputs, torch.Tensor):
+            outputs = [outputs]
+
+        # communicate meta outputs not real outputs for two reasons
+        # 1 - its faster (esp. since obj coll pickles tensor data!)
+        # 2 - avoid activating a cuda context for the src rank when unpickling on the recv end!
+        outputs_meta = tuple(
+            tree_map_only(torch.Tensor, lambda x: x.to("meta"), outputs)
         )
+        self._configure_outputs_meta(outputs_meta)
+
+        if (
+            self.is_last
+            or self.stage_index_to_group_rank[self.stage_index + 1] == self.group_rank
+        ):
+            logger.debug("%s skipping send to next stage", self.stage_index)
+
+        else:
+            logger.debug(
+                "%s sending to stage %s", self.stage_index, self.stage_index + 1
+            )
+            dist.send_object_list(
+                [outputs_meta],
+                dst=self.next_stage,
+                group=self.group,
+                device=self.device,
+            )
+            outputs_meta = tuple()
```

Comment:
Is there a specific need for this line?

Reply:
its clearing the value so we do not return it to the caller.

If i return this outputs_meta to the caller, i am asking them to feed it back as input to the next stage.  That's for the  V-schedule case where there is no send/recv comm between stages on the same rank at the V point. 

I'll add a comment to clarify.